analyzer/runtime/extract.go

```diff
@@ -727,6 +728,144 @@ func extractEvents(blockData *BlockData, relatedAccountAddresses map[apiTypes.Ad
+							Name:    "tokenID",
+							EvmType: "uint256",
+							// JSON supports encoding big integers, but many clients (javascript, jq, etc.)
+							// will incorrectly parse them as floats. So we encode uint256 as a string instead.
+							Value: tokenID.String(),
+						},
+					}
+					return nil
+				},
+				ERC721Approval: func(ownerEthAddr []byte, approvedEthAddr []byte, tokenIDU256 []byte) error {
+					if !bytes.Equal(ownerEthAddr, uncategorized.ZeroEthAddr) {
+						ownerAddr, err2 := registerRelatedEthAddress(blockData.AddressPreimages, relatedAccountAddresses, ownerEthAddr)
+						if err2 != nil {
+							return fmt.Errorf("owner: %w", err2)
+						}
+						eventData.RelatedAddresses[ownerAddr] = true
+					}
+					if !bytes.Equal(approvedEthAddr, uncategorized.ZeroEthAddr) {
+						approvedAddr, err2 := registerRelatedEthAddress(blockData.AddressPreimages, relatedAccountAddresses, approvedEthAddr)
+						if err2 != nil {
+							return fmt.Errorf("approved: %w", err2)
+						}
+						eventData.RelatedAddresses[approvedAddr] = true
+					}
+					if _, ok := blockData.PossibleTokens[eventAddr]; !ok {
+						blockData.PossibleTokens[eventAddr] = &evm.EVMPossibleToken{}
+					}
+					tokenID := &big.Int{}
+					tokenID.SetBytes(tokenIDU256)
+					eventData.EvmLogName = evmabi.ERC721.Events["Approval"].Name
+					eventData.EvmLogSignature = ethCommon.BytesToHash(event.Topics[0])
+					eventData.EvmLogParams = []*apiTypes.EvmEventParam{
+						{
+							Name:    "owner",
+							EvmType: "address",
+							Value:   ethCommon.BytesToAddress(ownerEthAddr),
+						},
+						{
+							Name:    "approved",
+							EvmType: "address",
+							Value:   ethCommon.BytesToAddress(approvedEthAddr),
+						},
+						{
+							Name:    "tokenID",
+							EvmType: "uint256",
+							// JSON supports encoding big integers, but many clients (javascript, jq, etc.)
+							// will incorrectly parse them as floats. So we encode uint256 as a string instead.
+							Value: tokenID.String(),
+						},
+					}
+					return nil
+				},
+				ERC721ApprovalForAll: func(ownerEthAddr []byte, operatorEthAddr []byte, approvedBool []byte) error {
+					if !bytes.Equal(ownerEthAddr, uncategorized.ZeroEthAddr) {
+						ownerAddr, err2 := registerRelatedEthAddress(blockData.AddressPreimages, relatedAccountAddresses, ownerEthAddr)
+						if err2 != nil {
+							return fmt.Errorf("owner: %w", err2)
+						}
+						eventData.RelatedAddresses[ownerAddr] = true
+					}
+					if !bytes.Equal(operatorEthAddr, uncategorized.ZeroEthAddr) {
+						operatorAddr, err2 := registerRelatedEthAddress(blockData.AddressPreimages, relatedAccountAddresses, operatorEthAddr)
+						if err2 != nil {
+							return fmt.Errorf("operator: %w", err2)
+						}
+						eventData.RelatedAddresses[operatorAddr] = true
+					}
+					if _, ok := blockData.PossibleTokens[eventAddr]; !ok {
+						blockData.PossibleTokens[eventAddr] = &evm.EVMPossibleToken{}
+					}
+					approved := approvedBool[31] != 0
```

Comment:
these topic+data parsing are all very ad-hoc, I'm hoping to do something easier in #421 with the go-ethereum libraries

Reply:
:+1: 
I expect that eventually, all this ERC-20 and ERC-721 handling from extract.go will move into a separate analyzer (that will use #421). Because right now, we trigger e.g. the ERC721Transfer handler just based on the fact that the event signature matches the ABI of ERC721 Transfer. We do not look at the token, and what we've learned about it.
Hrm, that will make dead reckoning trickier.
Anyway, not really important for this PR.

Reply:
> all this ERC-20 and ERC-721 handling from extract.go will move into a separate analyzer

I think not, this reckoning is special, which we won't do in general for other interfaces

Reply:
keeping it ad-hoc, marking as resolved